### ISSUE TITLE
[clang-tidy] False positive diagnostic for NewDeleteLeaks

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#122405]
Reproducer:

```C++
#include <vector>
#include <memory>

void RuleParseCommon() { std::vector{std::make_shared<int>()}; }
```

Reproducer reduced by cvise:

```C++
namespace std {
template <class E> struct initializer_list {
  const E *begin;
  const E *end;
};
template <typename _Tp> struct vector {
  vector(initializer_list<_Tp>);
};
struct _Sp_counted_base {};
struct __shared_count {
  __shared_count(int, int) {
    auto __pi = new _Sp_counted_base;
    _M_pi = __pi;
  }
  ~__shared_count();
  _Sp_counted_base *_M_pi;
};
int _M_ptr;
struct __shared_ptr {
  template <typename _Alloc>
  __shared_ptr(_Alloc __tag) : _M_refcount(_M_ptr, __tag) {}
  __shared_count _M_refcount;
};
struct shared_ptr : __shared_ptr {
  shared_ptr(int __tag) : __shared_ptr(__tag) {}
};
int allocate_shared___a;
shared_ptr allocate_shared() { return allocate_shared___a; }
shared_ptr make_shared() { return allocate_shared(); }
} // namespace std
using std::vector;

void RuleParseCommon() { vector{std::make_shared()}; }
```

Command:

```
clang-tidy test.cc  -checks='clang-analyzer-cplusplus.NewDeleteLeaks' -- -std=gnu++20
```

we got:

```
Potential leak of memory pointed to by field '_M_pi' [clang-analyzer-cplusplus.NewDeleteLeaks]
```

**Comment by steakhal** [comment]
I could not reproduce the issue: https://compiler-explorer.com/z/oMWdPW66a
What clang-tidy version do you use?
Could you reproduce the issue on Compiler Explorer?

**Comment by steakhal** [comment]
FYI I had a look at the preprocessed reduced version, and that of course reproduces the issue. However, without seeing the original case it's hard to tell if it's because the reduction actually implanted the leak due to the cuts it made or because it originally had the leak.

**Comment by steakhal** [comment]
The original code does not reproduce the FP on clang-16, clang-17, clang-18, clang-19, nor on clang-20.
```c++
#include <vector>
#include <memory>

void RuleParseCommon() { std::vector{std::make_shared<int>()}; } // <-- no warning
```

**Comment by ChuanqiXu9** [comment]
Then I am not sure if it relates to the standard library.  The reproducer will produce the FP consistent for me. My standard library is libstdc++9.

**Comment by steakhal** [comment]
Send the preprocessed file then.

**Comment by ChuanqiXu9** [comment]
> Send the preprocessed file then.

Sent it here: https://godbolt.org/z/cb1M6ozjE

**Comment by steakhal** [comment]
> > Send the preprocessed file then.
> 
> Sent it here: https://godbolt.org/z/cb1M6ozjE

Ill have a look at the case. Thanks!

**Comment by ChuanqiXu9** [comment]
> > > Send the preprocessed file then.
> > 
> > 
> > Sent it here: https://godbolt.org/z/cb1M6ozjE
> 
> Ill have a look at the case. Thanks!

Should we open the issue?
